base/pkg/read.jl

```diff
@@ -2,15 +2,70 @@
 
 module Read
 
-import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir
+import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir, ..isdevmetadata
 using ..Types
 
+devmetapkgpath(s::AbstractString) = joinpath(uppercase(string(s[1])), s * ".versions")
+
 readstrip(path...) = strip(readstring(joinpath(path...)))
 
-url(pkg::AbstractString) = readstrip(Dir.path("METADATA"), pkg, "url")
-sha1(pkg::AbstractString, ver::VersionNumber) = readstrip(Dir.path("METADATA"), pkg, "versions", string(ver), "sha1")
+function url(pkg::AbstractString, dev::Bool = false)
+    if dev
+        chomp(readline(joinpath(Dir.path("METADATA"), devmetapkgpath(pkg))))
+    else
+        readstrip(Dir.path("METADATA"), pkg, "url")
+    end
+end
+
+# Should only be called when cache is guaranteed to be updated
```

Comment:
I agree that this shouldn't be in this PR and the reason it is is that I simply forgot about non commited changes and I will update the PR to remove it The discussion here for me is what constraints it is worth to put on the interaction with METADATA in order to enable better caching. Allowing the user to muck around with arbitrary METADATA files while begin in an active julia session is inherently racy. In my view, the METADATA state could be cached at the first `Pkg` function and only `Pkg` functions and `PkgDev` are supported to edit METADATA while being in a julia session.